File: advent-of-code/2020/Day19.kt
sealed class DfaNode {

    abstract fun match(string: String): List<String>

    protected fun match(strings: List<String>): List<String> = strings.asSequence().flatMap(::match).distinct().toList()

    protected fun List<DfaNode>.match(string: String) = fold(listOf(string)) { acc, node -> node.match(acc) }
}

data class Leaf(val char: Char) : DfaNode() {

    override fun match(string: String): List<String> =
        if (!string.startsWith(char)) emptyList() else listOf(string.drop(1))
}

data class Node(val rules: MutableList<List<DfaNode>> = mutableListOf()) : DfaNode() {

    override fun match(string: String): List<String> = rules.flatMap { it.match(string) }
}

fun dfa(rules: Map<Int, String>, current: Int = 0, nodes: MutableMap<Int, DfaNode> = mutableMapOf()): DfaNode {
    var node = nodes[current]
    if (node != null) {
        return node
    }
    val rule = rules[current]!!

    if ('"' in rule) {
        node = Leaf(rule.replace("\"", "").singleOrNull().expect())
        nodes[current] = node
        return node
    }

    node = Node()
    nodes[current] = node
    node.rules.addAll(rule.split(" | ").map { it.split(' ').map { dfa(rules, it.toInt(), nodes) } })
    return node
}

fun countValid(rules: Map<Int, String>, messages: List<String>): Int {
    val root = dfa(rules)
    return messages.count { root.match(it).contains("") }
}

fun main() {
    val (ruleStrings, messages) = readBlocks().map { it.split('\n') }
    var rules = ruleStrings.map {
            val (number, restrictions) = it.split(": ")
            number.toInt() to restrictions
        }
        .toMap()

    val first = countValid(rules, messages)
    val second = countValid(rules + mapOf(8 to "42 | 42 8", 11 to "42 31 | 42 11 31"), messages)
    println("$first $second")
}

File: advent-of-code/2020/Day19.rb
class DfaNode
    def match(string)
        raise "abstract lol"
    end

    def matchList(strings)
        strings.flat_map {|it| match(it) }.uniq
    end
end

def listMatchList(nodes, string)
    list = [string]
    nodes.each {|x| list = x.matchList(list) }
    return list
end

class Leaf < DfaNode
    def initialize(char)
        @char = char
    end

    def match(string)
        if string.length == 0 || string[0] != @char
            []
        else
            string[1..]
        end
    end
end

class Node < DfaNode
    def initialize(rules)
        @rules = rules
    end

    def match(string)
        @rules.flat_map {|x| listMatchList(x, string) }
    end

    def add(rules)
        for i in rules
            @rules.append(i)
        end
    end
end

def dfa(rules, current, nodes)
    if nodes.include? current
        return nodes[current]
    end

    rule = rules[current]

    if rule.include? '"'
        node = Leaf.new rule[1]
        nodes[current] = node
        return node
    end

    node = Node.new []
    nodes[current] = node
    node.add(rule.split(" | ").map {|x| x.split(' ').map {|y| dfa(rules, y.to_i, nodes)} })
    return node
end

def count_valid(rules, messages)
    root = dfa(rules, 0, {})
    messages.filter {|x| root.match(x).include? "" }.length
end

input = $stdin.read.split("\n\n").map {|x| x.split("\n")}
rules = input[0].map {|x|
    splitted = x.split(": ")
    [splitted[0].to_i, splitted[1]]
}
messages = input[1]
first = count_valid(rules.to_h, messages)
rules.append([8, "42 | 42 8"], [11, "42 31 | 42 11 31"])
second = count_valid(rules.to_h, messages)
puts "#{first} #{second}"

